src/AppInstallerRepositoryCore/InstallerMetadataCollectionContext.cpp

```diff
@@ -0,0 +1,900 @@
+
+        HistoricalMetadataList.pop_back();
+        return true;
+    }
+
+    InstallerMetadataCollectionContext::InstallerMetadataCollectionContext() :
+        m_correlationData(std::make_unique<Correlation::ARPCorrelationData>())
+    {}
+
+    InstallerMetadataCollectionContext::InstallerMetadataCollectionContext(std::unique_ptr<Correlation::ARPCorrelationData> correlationData, const std::wstring& json) :
+        m_correlationData(std::move(correlationData))
+    {
+        auto threadGlobalsLifetime = InitializeLogging({});
+        InitializePreinstallState(json);
+    }
+
+    std::unique_ptr<InstallerMetadataCollectionContext> InstallerMetadataCollectionContext::FromFile(const std::filesystem::path& file, const std::filesystem::path& logFile)
+    {
+        THROW_HR_IF(E_INVALIDARG, file.empty());
+        THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND), !std::filesystem::exists(file));
+
+        std::unique_ptr<InstallerMetadataCollectionContext> result = std::make_unique<InstallerMetadataCollectionContext>();
+        auto threadGlobalsLifetime = result->InitializeLogging(logFile);
+
+        AICLI_LOG(Repo, Info, << "Opening InstallerMetadataCollectionContext input file: " << file);
+        std::ifstream fileStream{ file };
+
+        result->InitializePreinstallState(ConvertToUTF16(ReadEntireStream(fileStream)));
+
+        return result;
+    }
+
+    std::unique_ptr<InstallerMetadataCollectionContext> InstallerMetadataCollectionContext::FromURI(std::wstring_view uri, const std::filesystem::path& logFile)
+    {
+        THROW_HR_IF(E_INVALIDARG, uri.empty());
+
+        std::unique_ptr<InstallerMetadataCollectionContext> result = std::make_unique<InstallerMetadataCollectionContext>();
+        auto threadGlobalsLifetime = result->InitializeLogging(logFile);
+
+        std::string utf8Uri = ConvertToUTF8(uri);
+        THROW_HR_IF(E_INVALIDARG, !IsUrlRemote(utf8Uri));
+
+        AICLI_LOG(Repo, Info, << "Downloading InstallerMetadataCollectionContext input file: " << utf8Uri);
+
+        std::ostringstream jsonStream;
+        ProgressCallback emptyCallback;
+
+        const int MaxRetryCount = 2;
+        for (int retryCount = 0; retryCount < MaxRetryCount; ++retryCount)
+        {
+            bool success = false;
+            try
+            {
+                auto downloadHash = DownloadToStream(utf8Uri, jsonStream, DownloadType::InstallerMetadataCollectionInput, emptyCallback);
+
+                success = true;
+            }
+            catch (...)
+            {
+                if (retryCount < MaxRetryCount - 1)
+                {
+                    AICLI_LOG(Repo, Info, << "  Downloading InstallerMetadataCollectionContext input failed, waiting a bit and retrying...");
+                    Sleep(500);
+                }
+                else
+                {
+                    throw;
+                }
+            }
+
+            if (success)
+            {
+                break;
+            }

```

Comment:
Does breaking instead of setting `success = true` not work?

Reply:
Changed it in the code I copied from as well.